colossalai/tensor/shape_consistency.py

```diff
@@ -403,6 +405,157 @@ def get_all_one_step_transform_spec(self, source_spec: ShardingSpec, orig_cost_d
         valid_spec_dict.update(self.get_all_shard_spec(source_spec, orig_cost_dict))
         return valid_spec_dict
 
+    def mem_cost(self, comm_action_sequence: List[CommSpec]) -> TrainCycleItem:
+        """memory cost of the communication action sequence
+        TODO: Currently we just consider tensor numel in the shape consistency manger,
+        as the manager itself doesn't have the access to tensor dtype, we need to take
+        it into consideration in memory estimation.
+
+        Args:
+            comm_action_sequence (List[CommSpec]): list of communication actions
+
+        Returns:
+            TrainCycleItem: memory cost of such comm_action_sequence
+        """
+
+        def compute_shape(sharding_spec: ShardingSpec):
+            shape = sharding_spec.entire_shape
+            for dim, shard in sharding_spec.dim_partition_dict.items():
+                shape[dim] = shape[dim] // len(shard)
+            return shape
+
+        def gather_analysis(comm_spec: CommSpec, discard_input: bool, alloc_mem: int, peak_mem: int):
+            """analyze all_gather memory footprint
+            all_gather will allocate memory for the output tensor, and there will be temp memory for
+            all_gather operation, which is twice the size of output tensor
+
+            Args:
+                comm_spec (CommSpec): input CommSpec
+                discard_input (bool): whether to discard the input tensor
+                alloc_mem (int): current allocated memory
+                peak_mem (int): current peak memory
+            """
+            input_shape = compute_shape(comm_spec.sharding_spec)
+            input_numel = np.prod(input_shape)
+            output_numel = input_numel * comm_spec.device_mesh.mesh_shape[comm_spec.logical_process_axis]
+            peak_mem = max(peak_mem, alloc_mem + output_numel * 2)
+            alloc_mem += output_numel
+            if discard_input:
+                alloc_mem -= input_numel
+
+        def split_analysis(comm_spec: CommSpec, discard_input: bool, alloc_mem: int, peak_mem: int):
+            """analyze split memory footprint
+            split will allocate memory for the output tensor if we apply shard on the first dimension of
```

Comment:
If we apply shard on first dimension, no extra memory will be allocated?

Reply:
Yes, as we call `torch.narrow` in the `_split` action, it simply creates a reference to that part of the tensor, and as that part of the tensor is contiguous if we apply shard on the first dimension, the `.contiguous()` will not allocate any extra memory here.

Reply:
I mean the description in doc string is not correct. You should check it out.

Reply:
lol my bad